platooning_strategic/src/platoon_strategic.cpp

```diff
@@ -18,158 +18,350 @@
+    
 
 
 
+    int PlatoonStrategicPlugin::findLaneletIndexFromPath(int target_id, lanelet::routing::LaneletPath& path)
+    {
+        for(size_t i = 0; i < path.size(); ++i)
+        {
+            if(path[i].id() == target_id)
+            {
+                return i;
+            }
+        }
+        return -1;
     }
 
-    void PlatoonStrategicPlugin::run_states(){
-        
-        switch (psm_.current_platoon_state)
+    void PlatoonStrategicPlugin::bsm_cb(const cav_msgs::BSMConstPtr& msg)
+    {
+        cav_msgs::BSMCoreData bsm_core_ = msg->core_data;
+        host_bsm_id_ = bsmIDtoString(bsm_core_);
+    }
+
+    void PlatoonStrategicPlugin::pose_cb(const geometry_msgs::PoseStampedConstPtr& msg)
+    {
+        pose_msg_ = geometry_msgs::PoseStamped(*msg.get());
+
+        if (pm_.current_platoon_state != PlatoonState::STANDBY)
         {
-        case PlatoonState::STANDBY:
-            break;
-        case PlatoonState::LEADERWAITING:
-            waitingStartTime = ros::Time::now().toSec()*1000;
-            run_leader_waiting();
-            break;
+            lanelet::BasicPoint2d current_loc(pose_msg_.pose.position.x, pose_msg_.pose.position.y);
+            carma_wm::TrackPos tc = wm_->routeTrackPos(current_loc);
+            current_downtrack_ = tc.downtrack;
+            ROS_DEBUG_STREAM("current_downtrack_ = " << current_downtrack_);
+            current_crosstrack_ = tc.crosstrack;
+            ROS_DEBUG_STREAM("current_crosstrack_ = " << current_crosstrack_);
 
-        case PlatoonState::LEADER:
-            run_leader();
-            break;
+            pose_ecef_point_ = pose_to_ecef(pose_msg_);
+        }
         
-        case PlatoonState::CANDIDATEFOLLOWER:
-            candidatestateStartTime = ros::Time::now().toSec()*1000;
-            run_candidate_follower();
-            break;
         
-        case PlatoonState::FOLLOWER:
-            run_follower();
-            break;
-        
-        default:
-            throw std::invalid_argument("Invalid State");
-            break;
-        }
 
+    }
 
+    void PlatoonStrategicPlugin::cmd_cb(const geometry_msgs::TwistStampedConstPtr& msg)
+    {
+        cmd_speed_ = msg->twist.linear.x;
     }
 
-    void PlatoonStrategicPlugin::run()
+    void PlatoonStrategicPlugin::twist_cb(const geometry_msgs::TwistStampedConstPtr& msg)
     {
-    	initialize();
-        psm_.onMobilityRequestMessage(mobility_req_msg_);
-        psm_.onMobilityOperationMessage(mobility_op_msg_);
-        psm_.onMobilityResponseMessage(mobility_resp_msg_);
-        run_states();
-        ros::CARMANodeHandle::spin();
+        current_speed_ = msg->twist.linear.x;
+        if (current_speed_ < 0.01)
+        {
+            current_speed_ = 0.0;
+        }
+    }
 
+    void PlatoonStrategicPlugin::updateCurrentStatus(cav_msgs::Maneuver maneuver, double& speed, double& current_progress, int& lane_id){
+        if(maneuver.type == cav_msgs::Maneuver::LANE_FOLLOWING){
```

Comment:
Does it matter what these values are outside of the LANE_FOLLOWING maneuver? If we do LANE_FOLLOWING then a lane change, then more LANE_FOLLOWING, does the initial value of these variables impact the behavior?

Reply:
This plugin only generates lane following so this is just an extra check.